### ISSUE TITLE
🐛 slow linter on apply

### ISSUE THREAD

**Issue by nyanrus** [biomejs/biome#1569]
### Environment information

```block
CLI:
  Version:                      1.5.2
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           windows

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v20.9.0"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "pnpm/8.14.1"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 true

Workspace:
  Open Documents:               0
```

### What happened?

1. clone https://github.com/nyanrus/Floorp-core (originally cloned from Floorp-Projects/Floorp-core)
2. run
```block
pnpm add --save-dev --save-exact @biomejs/biome
pnpm dlx @biomejs/biome init
pnpm biome lint ./browser --apply
```
3. linter is fast, but applying is too slow (1891ms vs 147s)

### Expected result

the applying of the linter should be faster

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by nyanrus** [comment]
may related to #98

**Comment by nyanrus** [comment]
It seems linter filesystem problem because the `format --write` is fast enough
and linter without `--apply` is fast too.

**Comment by ematipico** [comment]
> We're also seeing some speed issues with `1.5.2` when running `biome check --apply .`. On `1.4.1` the command takes ~140 ms, but on `1.5.2` it takes ~1800 ms. Same amount of files, 497, listed as checked.
> 

Are you able to give us more information using debug logging? `--log-level=debug`

> Also, [`biome-vscode`](https://github.com/biomejs/biome-vscode) seems to still be running `1.4.1`, so there are some mismatches between the CLI and the results in VS Code.

Try to restart the LSP server; that's possibly the reason.

**Comment by ematipico** [comment]
> It seems linter filesystem problem because the `format --write` is fast enough and linter without `--apply` is fast too.

The issue is due to a performance bottleneck when we apply the code actions to big files. We haven't been able to allocate enough time to understand and fix this, unfortunately.

**Comment by believer** [comment]
> Are you able to give us more information using debug logging? `--log-level=debug`

The most notable thing I can see from the debug logging is that there's a bunch of `Unknown RomePath` logs, even from folders that we have set in `files.ignore`:

```
 2024-01-16T11:31:42.720218Z DEBUG  File capabilities: Unknown RomePath { path: "./dist/apps/api/src/lib/index.js.map" }
    at crates/biome_service/src/workspace/server.rs:90 on biome::worker_1
```

Our `biome.json` looks something like this:

```json
{
  "$schema": "https://biomejs.dev/schemas/1.5.2/schema.json",
  "files": {
    "ignore": [
      ".vscode/**",
      "android/**",
      "ios/**",
      "dist/**"
    ],
    "ignoreUnknown": true
  },
  "formatter": {
    "indentStyle": "space"
  },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true
    }
  },
  "organizeImports": {
    "enabled": true
  },
  "javascript": {
    "formatter": {
      "quoteStyle": "single",
      "semicolons": "asNeeded",
      "arrowParentheses": "always",
      "trailingComma": "es5"
    }
  },
  "overrides": [
    {
      "include": ["apps/app/**"],
      "linter": {
        "rules": {
          "suspicious": {
            "noConsoleLog": "error"
          },
          "style": {
            "useBlockStatements": "error"
          }
        }
      }
    },
    {
      "include": [
        "apps/api/**",
        "libs/data-aggregator/**"
      ],
      "linter": {
        "rules": {
          "suspicious": {
            "noConsoleLog": "off"
          },
          "complexity": {
            "noForEach": "off"
          }
        }
      }
    }
  ]
}
```

If we remove the `overrides`, `biome check --apply .` runs in about half the time listed above.

> Try to restart the LSP server; that's possibly the reason.

Ahh, of course it should use the local version first! [This PR to update](https://github.com/biomejs/biome-vscode/pull/82) confused us.

**Comment by si14** [comment]
Seeing a potentially related problem on 1.5.3 with [a large file](https://github.com/tldraw/tldraw/blob/34673156c4a719dc9832c6226669bdbcf674906d/packages/editor/src/lib/editor/Editor.ts) in vscode. Direct reformatting requests are instant, but saves with import formatting take a few seconds.

Test editor configured like this:
```
    "editor.codeActionsOnSave": {
        "source.organizeImports.biome": "always"
    },
```

Biome configured [like this](https://github.com/tldraw/tldraw/blob/34673156c4a719dc9832c6226669bdbcf674906d/biome.json)

Disabling import formatting with
```
	"organizeImports": {
		"enabled": false
	},
```
…makes saves instant.

I tried enabling logs for the extensions, here's the relevant bit that made me think it's related to https://github.com/biomejs/biome/issues/1569#issuecomment-1893436848 :
```
[Trace - 17:11:44] Sending notification 'textDocument/didChange'.
[Trace - 17:11:44] Received notification 'textDocument/publishDiagnostics'.
[Trace - 17:11:45] Sending request 'textDocument/codeAction - (5)'.
[Trace - 17:11:48] Received response 'textDocument/codeAction - (5)' in 3542ms.
[Trace - 17:11:48] Sending request 'textDocument/formatting - (6)'.
[Trace - 17:11:48] Received response 'textDocument/formatting - (6)' in 24ms.
[Trace - 17:11:48] Sending notification 'textDocument/didChange'.
[Trace - 17:11:48] Received notification 'textDocument/publishDiagnostics'.
[Trace - 17:11:48] Sending notification 'textDocument/didSave'.
```

**Comment by believer** [comment]
This issue has been resolved for us in `1.6.0`. Might be due to the new behavior of "Don't process files under an ignored directory."
